README.md
BibTeX-Ruby
===========

BibTeX-Ruby is the Rubyist's swiss-army-knife for all things BibTeX. It
includes a parser for all common BibTeX objects (@string, @preamble, @comment
and regular entries) and a sophisticated name parser that tokenizes correctly
formatted names; BibTeX-Ruby recognizes BibTeX string replacements, joins
values containing multiple strings or variables, supports cross-references,
and decodes common LaTeX formatting instructions to unicode; if you are in a
hurry, it also allows for easy export/conversion to formats such as YAML,
JSON, CiteProc/CSL, XML (BibTeXML, requires `rexml`), and RDF (experimental).

For a list of projects using BibTeX-Ruby, take a look at the
[project wiki](https://github.com/inukshuk/bibtex-ruby/wiki/Projects-Using-BibTeX-Ruby).


Quickstart
----------
Install and load BibTeX-Ruby in an IRB session:

    $ [sudo] gem install bibtex-ruby
    $ irb
    >> require 'bibtex'

Open a BibTeX bibliography:

```ruby
b = BibTeX.open('./ruby.bib')
```

Select a BibTeX entry and access individual fields:

```ruby
b['pickaxe'].title
#=> "Programming Ruby 1.9: The Pragmatic Programmer's Guide"
b[:pickaxe].author.length
#=> 3
b[:pickaxe].author.to_s
#=> "Thomas, D. and Fowler, Chad and Hunt, Andy"
b[:pickaxe].author[2].first
#=> "Andy"
```

Query a bibliography:

```ruby
b['@book'].length
#=> 3 - the number books in the bibliography
b['@article'].length
#=> 0 - the number of articles in the bibliography
b['@book[year=2009]'].length
#=> 1 - the number of books published in 2009
```

Extend first name initials throughout your bibliography:

```ruby
b.extend_initials ['Dave', 'Thomas']
b[:pickaxe].author.to_s
#=> "Thomas, Dave and Fowler, Chad and Hunt, Andy"
```

You can also extend all names in the bibliography to their prototypical
(i.e., the longest available) form:

```ruby
b.extend_initials! #=> extends all names in the bibliography
```

Use with caution as this method will treat two names as identical if they
look the same in their `#sort_order(:initials => true)` form.

Unify certain fields across the bibliography:

```ruby
b.unify :publisher, /o'?reilly/i, "O'Reilly"

b.unify :publisher, /^penguin/i do |entry|
  entry.publisher = 'Penguin Books'
  entry.address = 'London'
end
```

This will unify various spellings of entries published by O'Reilly and Penguin.

Render your bibliography in one of
[many different citation styles](https://github.com/citation-style-language/styles)
(requires the **citeproc-ruby** gem):

```ruby
require 'citeproc'
CiteProc.process b[:pickaxe].to_citeproc, :style => :apa
#=> "Thomas, D., Fowler, C., & Hunt, A. (2009). Programming Ruby 1.9: The Pragmatic Programmer's Guide. The Facets of Ruby. Raleigh, North Carolina: The Pragmatic Bookshelf."
CiteProc.process b[:pickaxe].to_citeproc, :style => 'chicago-author-date'
#=> "Thomas, Dave, Chad Fowler, and Andy Hunt. 2009. Programming Ruby 1.9: The Pragmatic Programmer's Guide. The Facets of Ruby.Raleigh, North Carolina: The Pragmatic Bookshelf."
CiteProc.process b[:pickaxe].to_citeproc, :style => :mla
#=> "Thomas, Dave, Chad Fowler, and Andy Hunt. Programming Ruby 1.9: The Pragmatic Programmer's Guide. Raleigh, North Carolina: The Pragmatic Bookshelf, 2009."
```

Save a bibliography to a file:

```ruby
b.save
#=> saves the original file
b.save_to(file)
#=> saves the bibliography in a new file
```

Compatibility
-------------
The BibTeX-Ruby gem has been developed and tested on Ruby 2.x, 1.9.3, and 1.8;
it has been confirmed to work with JRuby, Rubinius, and REE, however, there
have been repeated [issues](https://github.com/inukshuk/bibtex-ruby/issues)
(performance mostly) with MacRuby caused by MacRuby's current StringScanner
implementation.

Starting with BibTeX-Ruby version 3.0, support for Ruby versions 1.9.2 and
earlier has been dropped; most features will likely continue to work, but
compliance with old Rubies is not a priority going forward.

Starting with BibTex-Ruby version 7.0, only ruby versions still in "security maintenance phase" will be supported.
This is, as of writing Ruby 2.7 and up.

Documentation
-------------
It is very easy to use BibTeX-Ruby. You can use the top level utility methods
**BibTeX.open** and **BibTeX.parse** to open a '.bib' file or to parse a
string containing BibTeX contents. By default, BibTeX-Ruby will discard all
text outside of regular BibTeX elements; however, if you wish to include
everything, simply add `:include => [:meta_content]` to your invocation of
**BibTeX.open** or **BibTeX.parse**.

Once BibTeX-Ruby has parsed your '.bib' file, you can easily access individual
entries. For example, if you set up your bibliography as follows:

```ruby
b = BibTeX.parse <<-END
@book{pickaxe,
  address = {Raleigh, North Carolina},
  author = {Thomas, Dave and Fowler, Chad and Hunt, Andy},
  publisher = {The Pragmatic Bookshelf},
  series = {The Facets of Ruby},
  title = {Programming Ruby 1.9: The Pragmatic Programmer's Guide},
  year = {2009}
}
END
```

You could easily access it, using the entry's key, 'pickaxe', like so:
`b[:pickaxe]`; you also have easy access to individual fields, for example:
`b[:pickaxe][:author]`. Alternatively, BibTeX-Ruby accepts ghost methods to
conveniently access an entry's fields, similar to **ActiveRecord::Base**.
Therefore, it is equally possible to access the 'author' field above as
`b[:pickaxe].author`.

BibTeX-Ruby wraps all values of fields in an entry in Value
objects. This is necessary to transparently handle different types of values
(e.g., strings, dates, names etc.). These Value objects are designed to be
hardly discernible from regular Ruby strings, however, if you ever run into a
problem with a field's value, simply convert it to a string by calling the
`#to_s` method.

Instead of parsing strings you can also create BibTeX elements directly in
Ruby:

```ruby
bib = BibTeX::Bibliography.new
```

Using a Hash:

```ruby
bib << BibTeX::Entry.new({
  :bibtex_type => :book,
  :key => :rails,
  :address => 'Raleigh, North Carolina',
  :author => 'Ruby, Sam and Thomas, Dave, and Hansson, David Heinemeier',
  :booktitle => 'Agile Web Development with Rails',
  :edition => 'third',
  :keywords => 'ruby, rails',
  :publisher => 'The Pragmatic Bookshelf',
  :series => 'The Facets of Ruby',
  :title => 'Agile Web Development with Rails',
  :year => '2009'
})
```

Or programmatically:

```ruby
book = BibTeX::Entry.new
book.type = :book
book.key = :mybook
bib << book
```

### Cross References

From version 2.0, BibTeX-Ruby correctly resolves entry cross-references,
which are commonly used for entries with type `inbook`, `incollection`,
and `inproceedings`. When an entry has a valid citation key in the field
`crossref`, BibTeX-Ruby will return any fields inherited from the parent
entry:

```ruby
b = BibTeX.parse <<-END
@inbook{fraassen_1989b,
  Crossref = {fraassen_1989},
  Pages = {40-64},
  Title = {Ideal Science: David Lewis's Account of Laws},
}

@book{fraassen_1989,
  Address = {Oxford},
  Author = {Bas C. van Fraassen},
  Publisher = {Oxford University Press},
  Title = {Laws and Symmetry},
  Year = 1989
}
END
> b['fraassen_1989b'].booktitle
#=> <"Laws and Symmetry">
```

### Queries

Since version 1.3 BibTeX-Ruby implements a simple query language to search
Bibliographies via the `Bibliography#query` (or `Bibliography#q`) methods.
Additionally, you can access individual elements or groups of elements via
their index using `Bibliography#[]`; this accessor also exposes some of the
query functionality with the exception of yielding to a block. For instance:

```ruby
bib[-1]
#=> Returns the last element of the Bibliography or nil
bib[1,2]
#=> Returns the second and third elements or nil
bib[1..2]
#=> Same as above

bib[:key]
#=> Returns the first entry with key 'key' or nil
bib['key']
#=> Returns all entries with key 'key' or []

bib['@article']
#=> Returns all entries of type 'article' or []
bib['!@book']
#=> Returns all entries of any type other than 'book' or []
bib['@preamble']
#=> Returns all preamble objects (this is the same as Bibliography#preambles) or []
bib[/ruby/]
#=> Returns all objects that match 'ruby' anywhere or []

bib['@incollection[booktitle]']
#=> Returns all in-collection entries with a booktitle or []

# note that the above includes entries inheriting the book title
# from a cross-referenced entry!

bib['@book[keywords=ruby]']
#=> Returns all books whose keywords attribute equals 'ruby' or []
bib['@book[keywords!=ruby]']
#=> Returns all books whose keywords attribute does not equal 'ruby'
bib['@book[keywords/=ruby]']
#=> Same as above

bib.q('@book[keywords ^= ruby]')
#=> Returns all books whose keywords attribute matches /^ruby/
bib.q('@book[keywords ~= ruby]')
#=> Returns all books whose keywords attribute matches /ruby/
bib['@book[keywords!~ruby]']
#=> Returns all books whose keywords attribute does not match /ruby/ or don't have keywords attribute

bib.q('@article[year<=2007]')
#=> Returns all articles published in 2007 or earlier
bib.query('@book') { |e| e.keywords.split(/,/).length > 1 }
#=> Returns all book entries with two or more keywords or []
```

Queries offer syntactic sugar for common enumerator invocations:

```ruby
bib.query(:all, '@book')
#=> same as bib.select { |b| b.has_type?(:book) }
bib.query('@book')
#=> same as above
bib.query(:first, '@book')
#=> same as bib.detect { |b| b.has_type?(:book) }
bib.query(:none, '@book')
#=> same as bib.reject { |b| b.has_type?(:book) }
```

You can also use queries to delete entries in your bibliography:

```ruby
bib.delete(/ruby/)
#=> deletes all object that match 'ruby' in their string representation
bib.delete('@comment')
#=> strips all BibTeX comments from the bibliography
```

### String Replacement

If your bibliography contains BibTeX @string objects, you can let BibTeX-Ruby
replace the strings for you. You have access to a bibliography's strings via
**BibTeX::Bibliography#strings** or by using a '@string' query.
You can replace the string symbols of an object by calling the object's
the **replace** method. Thus, to replace all strings defined in bibliography
b you could use the following code:

```ruby
b.each do |obj|
  obj.replace(b.q('@string'))
end
```

A shorthand version for replacing all strings in a given bibliography is the
`Bibliography#replace` method. Similarly, you can use the
`Bibliography#join` method to join individual strings together. For instance:

```ruby
    bib = BibTeX::Bibliography.new
    bib.add BibTeX::Element.parse '@string{ foo = "foo" }'
    bib << BibTeX::Element.parse '@string{ bar = "bar" }'
    bib.add BibTeX::Element.parse <<-END
     @book{abook,
       author = foo # "Author",
       title = foo # bar
     }
    END
    puts bib[:abook].to_s
    #@book{abook,
    #  author = foo # "Author",
    #  title = foo # bar
    #}
    bib.replace
    puts bib[:abook].to_s
    #@book{abook,
    #  author = "foo" # "Author",
    #  title = "foo" # "bar"
    #}
    bib.join
    #@book{abook,
    #  author = {fooAuthor},
    #  title = {foobar}
    #}
```

### Names

Since version 1.3, BibTeX-Ruby features a name parser. You can use the
top-level `BibTeX.names` utility to quickly parse individual name values.
Alternatively, you can call `Bibliography.parse_names` to convert all name
fields contained in the bibliography. When parsing BibTeX files, BibTeX-Ruby
will automatically convert names; if you do not want the names to be parsed
you can set the `:parse_names` parser option to false.

Note that the string replacement and concatenation features described above
are not supported for name objects; therefore, BibTeX-Ruby tries to replace
and join all values before name conversion; name fields containing string
symbols that cannot be replaced will not be parsed.

In the following example, string replacement can take place, thus all names
are parsed and can easily be mapped to their last names:

```ruby
BibTeX.parse(<<-END)[1].author.map(&:last)
  @string{ ht = "Nathaniel Hawthorne" }
  @book{key,
    author = ht # " and Melville, Herman"
  }
  END
#=> ["Hawthorne", "Melville"]
```

Another useful method is `Bibliography#names` which returns all names in
your bibliography (authors, editors, translators). For example, to quickly
expand the initials of a name across your entire bibliography, you could
use the following snippet:

```ruby
b.names.each do |name|
  if name.sort_order =~ /^Poe, E/
    name.first = 'Edgar Allen'
  end
end
```

There is also a short-hand for this use case:

```ruby
b.extend_initials ['Edgar Allen', 'Poe']
```

Alternatively, if your bibliography contains the same names in various
forms (e.g., 'Poe, Edgar A.', 'Poe, E.A.', 'Poe, E. Allen') you can also
set all names to their longest available form:

```ruby
b.extend_initials!
```

Use with caution, though, as this method will treat names as identical
as long as their initials are the same. That is to say, 'Poe, Eric A.' would
be extend to 'Poe, Edgar Allen'.

### Duplicates

Large bibliographies often contain duplicate data, i.e., duplicate entries
which are not completely identical (e.g., authors or editors with first
names or initials, titles using different casing, different keywords etc.).
BibTex-Ruby allows you to group your bibliography by any number of fields
in order to detect such duplicate entries.

```ruby
b.select_duplicates_by :year, :title
#=> groups the bibliography by using the year and title field as key

b.duplicates?
#=> whether or not the bibliography contains any duplicates
```

For more complex requirements you can use the `#group_by` method directly.
This methods accepts a list of arguments whose value will be used for grouping
and, additionally, a block. The current digest and each individual entry will
be passed to the block and the block's return value is used as the final
digest.

The duplicate methods above, for example, do something like this:

```ruby
group_by(:year, :title) do |digest, entry|
  digest.gsub(/\s+/, '').downcase
end
```

You can use this method, for example, to match entries only by their author's
last name and so on and so forth.

Since version 4.1.0, BibTeX-ruby supports parsing of names spelled in the East Asian order (last name, then first name). Names contain a Chinese, Japanese or Korean letter are currently assumed to be such names. If the name is written in comma-separated manner, it is parsed in the normal way.

```ruby
b = BibTeX.parse(<<-END)
@book{key,
  title = "プログラミング言語 Ruby",
  author = "David Flanagan and まつもと ゆきひろ",
  translator = "卜部, 昌平 and 長尾, 高弘",
  year = "2009",
  publisher = "O'Reilly Japan"
}
END
[*b[0].author, *b[0].translator].map(&:last) #=> [Flanagan, まつもと, 卜部, 長尾]
```

### Filters

Since version 1.3.8 BibTeX-Ruby comes with a plugin framework for input
filters. You can use the methods `convert` and `convert!` methods if `Value`,
`Entry` and `Bibliography` instances to easily convert string values according
to a given filter. Starting with version 1.3.9 BibTeX-Ruby includes a
LaTeX filter that depends on the
[latex-decode gem](http://rubygems.org/gems/latex-decode). Example:

```ruby
faust = '@book{faust, title = {Faust: Der Trag\"odie Erster Teil}}'
BibTeX.parse(faust).convert(:latex)[:faust].title
#=> "Faust: Der Tragödie Erster Teil"
```

Conditional conversions are also supported:

```ruby
faust1 = '@book{faust1, title = {Faust: Der Trag\"odie Erster Teil}}'
faust2 = '@book{faust2, title = {Faust: Der Trag\"odie Zweiter Teil}}'
p BibTeX.parse(faust1 + faust2).convert(:latex) { |e| e.key == :faust2 }.to_s
```

Returns:

```bibtex
@book{faust1,
  title = {Faust: Der Trag\"odie Erster Teil}
}
@book{faust2,
  title = {Faust: Der Tragödie Zweiter Teil}
}
```

If you need to express a condition on the basis of individual fields, use the
conversion methods of BibTeX::Entry with a block instead (the block will be
passed the key and value of each field prior to conversion).

When working with Bibliographies that contain LaTeX it is often best to
apply the filter upon opening or parsing the Bibliography. You can do this,
by passing the `:filter` option:

```ruby
BibTeX.open 'references.bib', :filter => :latex
```


History.txt
* Dual-license bibtex-ruby under GPL-3.0 and BSD-2-Clause
* Add official support for Ruby 3.1
* Remove official support for Ruby < 2.7, as they reached their End-of-Life

6.0.0 / 2021-01-07
==================

* Make bibtexml converter optional

5.1.7 / 2021-01-07
==================

* Add support for Ruby 3.0 (@skalee)

5.1.6 / 2020-12-02
==================

* Ignore case when matching types
* Remove official support for Ruby 2.4

5.1.5 / 2020-10-14
==================

5.1.4 / 2020-04-14
==================

5.1.3 / 2020-04-06
==================

5.1.2 / 2020-02-27
==================

5.1.1 / 2020-01-17
==================

5.1.0 / 2020-01-17
==================

* Use File.read instead of Kernel.open in BibTeX.open
* Fix Ruby 2.7 deprecation warnings

5.0.0 / 2019-06-12
==================

* Dropped support for Ruby < 2.3

4.2.0 / 2015-02-02
==================

* Fixed queries being whitespace sensitive

4.0.9 / 2015-01-20
==================

* Improve BibTeX::Value comparison

4.0.8 / 2015-01-12
==================

* Ruby 2.2 support
* Added Entry#month_numeric
* Added support for numeric values

4.0.4 / 2014-10-10
==================

* added !~ matcher for queries (@temporaer)

4.0.0 / 2014-07-08
==================

* BREAKING: Support type/key fields. Rename type -> bibtex_type
  and key -> bibtex_key in key-value exports.
* Add support for BibLaTeX date syntax

3.1.6 / 2014-06-17
==================

* Allow semicolons in keys
* Bugfixes

3.1.5 / 2014-03-25
==================

* Improve Name#extend_initials
* Add Name#normalize_initials

3.1.4 / 2014-03-15
==================

* Add more name fields
* Improve CiteProc export

3.1.3 / 2014-02-10
==================

* Fix Ruby warnings

3.1.2 / 2014-01-29
==================

* Improve CiteProc export (thanks @andriusvelykis)

3.1.1 / 2014-01-20
==================

* RDF Export improvements (@tmaier)

3.1.0 / 2014-01-20
==================

* Fixed and refactored RDF Export (@tmaier)

3.0.1 / 2013-12-30
==================

* Fixed sorting (@casutton)

3.0.0 / 2013-11-11
==================

* Added transliteration support to default key generator (@cruessler)
* Improve parsing of non-ascii keys (@cruessler)

2.3.4 / 2013-08-22
==================

* Fixed file-permissions in gem (thanks to @jwvdm)

2.3.3 / 2013-07-29
==================

* Fixed #duplicates? method (thanks to @mfenner)
* Fixed issue with JSON export (thanks to @noahdesu)

2.3.1 / 2013-07-04
==================

* Support multiple filters in #convert methods

2.3.0 / 2013-06-04
==================

* Add attribute-only queries (thanks to @ashinkarov)
* Add negated queries (thanks to @ashinkarov)

2.2.2 / 2013-15-01
==================

* Improves CiteProc export #61 #62 (@plessl)

2.2.1 / 2013-13-01
==================

* Fixes BibTeX/CiteProc type issue (thanks to @plessl)

2.2.0 / 2012-12-02
==================

* Adds duplicate detection
* Adds Entry merge support

2.1.1 / 2012-11-02
==================

* Fixed lexer issue with whitespace before cite keys (@lmullen)

2.1.1 / 2012-09-14
==================

* Improved @comment and @preamble parsing (@JLimperg)

2.1.0 / 2012-08-31
==================

* Improved query support
* Use Travis-CI
* Updated dependencies
* Bug fixes and 1.8 compatibility (@RubenVerborgh)

2.0.11 / 2012-05-23
===================

* Added :allow_missing_keys option (thanks to @jrochkind)

2.0.9 / 2012-05-15
==================

* Added extend_initials! method to extend all names in the bibliography
* Added unify method as a shorthand to unify fields across the bibliography

2.0.8 / 2012-05-11
==================

* Added extend_initials feature for names
* Added experimental RDF export (using BIBO ontology)

2.0.7 / 2012-04-30
==================

* Added support for <=, >= numeric queries

2.0.6 / 2012-04-30
==================

* Added support for !=, /=, ~=, and ^= queries

2.0.5 / 2012-04-19
==================

* Fixed Entry#dup (@JLimperg)
* Implemented explicit caching of regular expressions (for MacRuby)
* Switched to new MultiJson API

2.0.4 / 2011-12-20
==================

* Added lexer option :strip to strip newlines from literals (thanks to @JLimperg)
* Added :linebreaks filter


2.0.3 / 2011-12-13
==================

* Added new entry wildcard "@*" to queries (kudos to @vanto)

2.0.2 / 2011-11-25
==================

* Fixed (#33): error in Entry#respond_to? (kudos to @houshuang)

2.0.1 / 2011-10-20
==================

* Fixed name parser incompatibility with Rails (kudos to @amattsmith)
* Improved latex-decoder (kudos to @etc)

2.0.0 / 2011-09-25
==================

* Added support to directly add Arrays, Hashes, or Strings to Bibliographies
* Added cross reference support (@etc)
* Moved back to using strings as Entry keys (instead of symbols)
* Added improved default key generation
* Improved name parsing
* Switched to multi_json gem for JSON and citeproc export
* Switched to BibTeXML as XML export
* Added duplicate detection
* Bugfixes

1.3.12 / 2011-09-06 
===================================

* Improved default keys for BibTeX entries

1.3.11 'happy #rbxday' / 2011-08-05 
===================================

* Rubinius compatibility
* Cleaned up dependency loading

1.3.10 / 2011-07-16 
==================

* Fixes gemspec options
* Added support for cite keys containing spaces (@sharnik) 

=== 1.3.9 / 2011-07-05

* Added a LaTeX input filter

=== 1.3.8 / 2011-07-05

* Added a converter-plugin framework

=== 1.3.7 / 2011-06-30

* Added option to toggle dropping/non-dropping behaviour of 'von' parts
  when exporting to CiteProc/CSL
* Fixed name parser not accepting 'von' parts starting with upper case words
  in sort order (#11)
* Added support for multi-line string literals (#15)

=== 1.3.6 / 2011-06-11

* Refactored lexer and parser for speed
* Added support for @strings with braced values (#10)

=== 1.3.5 / 2011-06-07

* Removed debugger initialisation

=== 1.3.4 / 2011-06-07

* Performance boost for parsing of large bibliographies


lib/bibtex/parser.rb
#
# DO NOT MODIFY!!!!
# This file is automatically generated by Racc 1.4.11
# from Racc grammer file "".
#

require 'racc/parser.rb'

require 'bibtex/lexer'

module BibTeX
  class Parser < Racc::Parser

module_eval(<<'...end bibtex.y/module_eval...', 'bibtex.y', 89)

  attr_reader :lexer, :options
  
  @defaults = {
    :include => [:errors],
    :allow_missing_keys => false,
    :debug => false
  }.freeze
  
  class << self
    attr_reader :defaults
  end
  
  def initialize(options = {})
    @options = Parser.defaults.merge(options)
    @lexer = Lexer.new(@options)
  end

  def parse(input)
    @yydebug = debug?   

    lexer.analyse(input)    
    do_parse
    #yyparse(@lexer,:each)
  end
  
  def next_token
    lexer.next_token
  end
  
  def debug?
    options[:debug] || ENV['DEBUG']
  end
  
  def allow_missing_keys?
    options[:allow_missing_keys]
  end
  
  def missing_key
    unless allow_missing_keys?
      raise ParseError, "Failed to parse BibTeX entry: cite-key missing"
    end
  end
  
  def on_error(tid, val, vstack)
    message =
      "Failed to parse BibTeX on value #{val.inspect} (#{token_to_str(tid) || '?'}) #{ vstack.inspect}"

    BibTeX.log.error message
    raise ParseError, message
  end

# -*- racc -*-
...end bibtex.y/module_eval...
##### State transition tables begin ###

racc_action_table = [
    14,    33,    38,    26,    32,    33,    24,    18,    32,    15,
    34,    39,    16,    37,    34,    33,    33,    36,    32,    32,
    52,    27,     4,     4,    34,    34,     6,     6,    26,     5,
     5,    48,    43,    44,    29,    41,    42,    22,    29,    46,
    47,    21,    20,    19,    54,     7,    44,    44 ]

racc_action_check = [
     4,    21,    23,    17,    21,    47,    17,     4,    47,     4,
    21,    26,     4,    23,    47,    44,    39,    22,    44,    39,
    39,    18,     0,     2,    44,    39,     0,     2,    38,     0,
     2,    38,    30,    30,    20,    27,    28,    16,    33,    35,
    36,    15,    14,     7,    45,     1,    50,    55 ]

racc_action_pointer = [
    20,    45,    21,   nil,    -4,   nil,   nil,    43,   nil,   nil,
   nil,   nil,   nil,   nil,    34,    33,    29,    -8,    13,   nil,
    29,    -7,     6,    -1,   nil,   nil,     4,    25,    22,   nil,
    18,   nil,   nil,    33,   nil,    25,    33,   nil,    17,     8,
   nil,   nil,   nil,   nil,     7,    30,   nil,    -3,   nil,   nil,
    31,   nil,   nil,   nil,   nil,    32 ]

racc_action_default = [
    -1,   -34,    -2,    -3,   -34,    -6,    -7,   -34,    -4,    -5,
    -8,    -9,   -10,   -11,   -34,   -34,   -34,   -34,   -34,    56,
   -13,   -34,   -34,   -34,   -25,   -29,   -34,   -27,   -34,   -14,
   -34,   -18,   -20,   -13,   -22,   -34,   -34,   -23,   -34,   -34,
   -26,   -28,   -12,   -15,   -34,   -34,   -16,   -34,   -24,   -30,
   -32,   -31,   -33,   -19,   -21,   -17 ]

racc_goto_table = [
    30,    25,    28,     3,     1,     8,    11,    12,    13,    10,
     9,    35,    53,    17,    23,    45,    40,     2,    50,    51,
   nil,   nil,    49,   nil,   nil,   nil,    55 ]

racc_goto_check = [
    10,    16,     9,     3,     1,     3,     6,     7,     8,     5,
     4,    11,    12,    13,    14,     9,    15,     2,    10,    17,
   nil,   nil,    16,   nil,   nil,   nil,    10 ]

racc_goto_pointer = [
   nil,     4,    17,     3,     6,     5,     2,     3,     4,   -18,
   -21,   -11,   -32,     9,    -3,   -11,   -16,   -20 ]

racc_goto_default = [
   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,   nil,
   nil,   nil,    31,   nil,   nil,   nil,   nil,   nil ]

racc_reduce_table = [
  0, 0, :racc_error,
  0, 19, :_reduce_1,
  1, 19, :_reduce_2,
  1, 20, :_reduce_3,
  2, 20, :_reduce_4,
  2, 21, :_reduce_5,
  1, 21, :_reduce_6,
  1, 21, :_reduce_7,
  1, 22, :_reduce_8,
  1, 22, :_reduce_9,
  1, 22, :_reduce_10,
  1, 22, :_reduce_11,
  4, 23, :_reduce_12,
  0, 27, :_reduce_13,
  1, 27, :_reduce_14,
  4, 25, :_reduce_15,
  4, 24, :_reduce_16,
  3, 29, :_reduce_17,
  1, 28, :_reduce_18,
  3, 28, :_reduce_19,
  1, 30, :_reduce_20,
  3, 30, :_reduce_21,
  1, 30, :_reduce_22,
  3, 26, :_reduce_23,
  4, 26, :_reduce_24,
  2, 26, :_reduce_25,
  3, 31, :_reduce_26,
  0, 33, :_reduce_27,
  1, 33, :_reduce_none,
  1, 32, :_reduce_29,
  3, 32, :_reduce_30,
  3, 34, :_reduce_31,
  1, 35, :_reduce_32,
  1, 35, :_reduce_33 ]

racc_reduce_n = 34

racc_shift_n = 56

racc_token_table = {
  false => 0,
  :error => 1,
  :AT => 2,
  :COMMA => 3,
  :COMMENT => 4,
  :CONTENT => 5,
  :ERROR => 6,
  :EQ => 7,
  :LBRACE => 8,
  :META_CONTENT => 9,
  :KEY => 10,
  :NAME => 11,
  :NUMBER => 12,
  :PREAMBLE => 13,
  :RBRACE => 14,
  :SHARP => 15,
  :STRING => 16,
  :STRING_LITERAL => 17 }

racc_nt_base = 18

racc_use_result_var = true

Racc_arg = [
  racc_action_table,
  racc_action_check,
  racc_action_default,
  racc_action_pointer,
  racc_goto_table,
  racc_goto_check,
  racc_goto_default,
  racc_goto_pointer,
  racc_nt_base,
  racc_reduce_table,
  racc_token_table,
  racc_shift_n,
  racc_reduce_n,
  racc_use_result_var ]

Racc_token_to_s_table = [
  "$end",
  "error",
  "AT",
  "COMMA",
  "COMMENT",
  "CONTENT",
  "ERROR",
  "EQ",
  "LBRACE",
  "META_CONTENT",
  "KEY",
  "NAME",
  "NUMBER",
  "PREAMBLE",
  "RBRACE",
  "SHARP",
  "STRING",
  "STRING_LITERAL",
  "$start",
  "bibliography",
  "objects",
  "object",
  "at_object",
  "comment",
  "string",
  "preamble",
  "entry",
  "content",
  "string_value",
  "string_assignment",
  "string_literal",
  "entry_head",
  "assignments",
  "opt_key",
  "assignment",
  "value" ]

Racc_debug_parser = false

##### State transition tables end #####

# reduce 0 omitted

module_eval(<<'.,.,', 'bibtex.y', 32)
  def _reduce_1(val, _values, result)
     result = BibTeX::Bibliography.new(@options) 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 33)
  def _reduce_2(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 35)
  def _reduce_3(val, _values, result)
     result = BibTeX::Bibliography.new(@options) << val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 36)
  def _reduce_4(val, _values, result)
     result << val[1] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 38)
  def _reduce_5(val, _values, result)
     result = val[1] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 39)
  def _reduce_6(val, _values, result)
     result = BibTeX::MetaContent.new(val[0]) 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 40)
  def _reduce_7(val, _values, result)
     result = BibTeX::Error.new(val[0]) 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 42)
  def _reduce_8(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 43)
  def _reduce_9(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 44)
  def _reduce_10(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 45)
  def _reduce_11(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 47)
  def _reduce_12(val, _values, result)
     result = BibTeX::Comment.new(val[2]) 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 49)
  def _reduce_13(val, _values, result)
     result = '' 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 50)
  def _reduce_14(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 52)
  def _reduce_15(val, _values, result)
     result = BibTeX::Preamble.new(val[2]) 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 54)
  def _reduce_16(val, _values, result)
     result = BibTeX::String.new(val[2][0],val[2][1]); 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 56)
  def _reduce_17(val, _values, result)
     result = [val[0].downcase.to_sym, val[2]] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 58)
  def _reduce_18(val, _values, result)
     result = [val[0]] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 59)
  def _reduce_19(val, _values, result)
     result << val[2] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 61)
  def _reduce_20(val, _values, result)
     result = val[0].downcase.to_sym 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 62)
  def _reduce_21(val, _values, result)
     result = val[1] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 63)
  def _reduce_22(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 65)
  def _reduce_23(val, _values, result)
     result = val[0] << val[1] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 66)
  def _reduce_24(val, _values, result)
     result = val[0] << val[1] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 67)
  def _reduce_25(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 69)
  def _reduce_26(val, _values, result)
     result = BibTeX::Entry.new(:bibtex_type => val[0].downcase.to_sym, :bibtex_key => val[2]) 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 71)
  def _reduce_27(val, _values, result)
     missing_key 
    result
  end
.,.,

# reduce 28 omitted

module_eval(<<'.,.,', 'bibtex.y', 74)
  def _reduce_29(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 75)
  def _reduce_30(val, _values, result)
     result.merge!(val[2]) 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 77)
  def _reduce_31(val, _values, result)
     result = { val[0].downcase.to_sym => val[2] } 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 79)
  def _reduce_32(val, _values, result)
     result = val[0] 
    result
  end
.,.,

module_eval(<<'.,.,', 'bibtex.y', 80)
  def _reduce_33(val, _values, result)
     result = val[0] 
    result
  end
.,.,

def _reduce_none(val, _values, result)
  val[0]
end

  end   # class Parser
  end   # module BibTeX
